_posts/2018-11-09-conditional-types-in-typescript.markdown

```diff
@@ -0,0 +1,402 @@
+hierarchy matter. The two aren't mutually exclusive, however. Some languages, like Scala and Flow, allow you to mix
+and match.
+
+Aside from that, the way that assignability works in TypeScript is quite intuitive.
+
+```ts
+interface Shape {
+  color: string
+}
+
+class Circle {
+  color: string
+  radius: number
+}
+
+// ✔ All good! Cirlces have a color
+const shape: Shape = new Circle()
+// ✘ Type error! Not all shapes have a radius!
+const circle: Circle = shape
+```
+
+Speaking structurally we can say that `A extends B` is a lot like '`A` is a superset of `B`', or, to be more
+verbose, '`A` has all of `B`'s properties, and maybe some more'.
+
+There's one minor caveat though, and that's with literal types. In TypeScript you can use literal values of
+primitive types as types themselves.
+
+```ts
+let fruit: "banana" = "banana"
+
+// Type Error! "apple" is not assignable to "banana"
+fruit = "apple"
+```
+
+The string `"banana"` doesn't have any more or fewer properties than any other `string`. But the type `"banana"`
+is, conceptually, more _specific_ than the type `string`.
+
+So another way to think of it is that `A extends B` is like '`A` is a possibly-more-specific version of `B`'.
+
+Which brings us to 'top' and 'bottom' types: the _least_ and _most_ specific types, respectively.
+
+In type theory a 'top' type is one which all other types are assignable to. It is the type you use to say "I have
+absolutely no information about what this value is". TypeScript has two top types: `any` and `unknown`.
+
+- Using `any` is like saying "I have no idea what this value looks like. So, TypeScript, please assume I'm using it
+  correctly, and don't complain if anything I do seems unsafe.".
+- Using `unknown` is like saying "I have no idea what this value looks like. So, TypeScript, please make sure I
+  check what it is capable of at run time."
+
+A 'bottom' type is one which no other types are assignable to, and that no values can be an instance of. TypeScript
+has one bottom type: `never`. That's a nice descriptive name because it literally means _this can never happen_.
+
+Top and bottom types are useful to know about when working with conditional types. `never` is especially useful
+when using conditional types to refine unions...
+
+## Refining unions with distributive conditional types
+
+Conditional types let you filter out particular members of a union type. To illustrate, let's say we have a union
+type called `Animal`:
+
+```ts
+type Animal = Lion | Zebra | Tiger | Shark
+```
+
+And imagine that we needed to write a function that used only those animals which are also cats. We might write
+some helper type called `ExtractCat` to do that:
+
+```ts
+type ExtractCat<A> = A extends { meow(): void } ? A : never
+
+type Cat = ExtractCat<Animal>
+// => Lion | Tiger
+```
+
+_I know lions and tigers don't meow, but how cute would it be if they did_ ^\_^
```

Comment:
Haha, nice note! The biology behind cat noises is super-interesting, actually: https://www.youtube.com/watch?v=T_yxZFjqQUg

Reply:
haha that's fascinating :)